### ISSUE TITLE
How to query if a field value is None

### ISSUE THREAD

**Issue by rmcolq** [CLIMB-TRE/onyx-client#45]
At the moment, none values are stored as "". Is it possible to add None as an alternative match to "" when doing filters/queries?

**Comment by tombch** [comment]
Yeah I'm not immediately sure how to handle this. There is a bit of a confusing system in that empty text fields are specified with `""`, whereas anything else (dates, numbers, true/false) are queried using `None`. Its because in the actual database, empty text fields are stored as the empty string, whereas empty anything-else are stored as database NULL values (which can be specified with None in python). At the very least, I need to document this

**Comment by tombch** [comment]
I should say, that for non-text fields, there is a `isnull` lookup that accepts a true/false value. But this doesn't work for text fields. So e.g. to get all empty collection_date values, you can do `onyx filter <project> --field collection_date.isnull=true`

**Comment by tombch** [comment]
Just to update to this: I am working on a unified `empty` lookup that will check for the empty string for text values, and for a NULL value otherwise. This should mean people will be able to do stuff like `onyx filter <project> --field collection_date.isempty=true --field text_field.isempty=false` etc

**Comment by warrickball** [comment]
I just had a quick look at this and it seems that Django is quite opinionated about not allowing null values for strings. Quoting the [Django fields docs](https://docs.djangoproject.com/en/5.0/ref/models/fields/):

> Avoid using [null](https://docs.djangoproject.com/en/5.0/ref/models/fields/#django.db.models.Field.null) on string-based fields such as [CharField](https://docs.djangoproject.com/en/5.0/ref/models/fields/#django.db.models.CharField) and [TextField](https://docs.djangoproject.com/en/5.0/ref/models/fields/#django.db.models.TextField). If a string-based field has null=True, that means it has two possible values for “no data”: NULL, and the empty string. In most cases, it’s redundant to have two possible values for “no data;” the Django convention is to use the empty string, not NULL.

I experimented a bit and I think searching on a text-based field with something like
```
$ onyx filter mscapetest --field biosample_source_id=""
```
should work.

From a development perspective, I'd suggest simply inheriting this quirk of Django's defaults and not trying to work around it, though I agree it needs to be documented properly, which I can follow up on. If a workaround like the `empty` lookup you suggest happens to be straightforward, then go ahead.

The other possibility is that you can set the Django parameters (e.g. `Field.null=True`, maybe) so that it will accept both the empty string *or* null but as the docs suggest, the risk is that the database ends up with a mix of empty strings and null values and searches for one miss the other, etc.

**Comment by tombch** [comment]
This issue is now solved! With the latest version of the client, `v4.5.0`, things like:
```python
import os
from onyx import OnyxClient, OnyxEnv, OnyxConfig

config = OnyxConfig(
    domain=os.environ[OnyxEnv.DOMAIN],
    token=os.environ[OnyxEnv.TOKEN],
)
client = OnyxClient(config)

# Get all records with empty specimen_type_details
records = list(client.filter("mscape", specimen_type_details=None))
```
can now be done and should work consistently across different types of fields.
